src/vm/comutilnative.cpp

```diff
@@ -2624,14 +2623,167 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
 }
 FCIMPLEND
 
-// Return true if the valuetype does not contain pointer and is tightly packed
+static BOOL HasOverriddenMethod(MethodTable* mt, MethodTable* classMT, WORD methodSlot)
+{
+    CONTRACTL{
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+        SO_TOLERANT;
+    } CONTRACTL_END;
+
+    PCODE actual = mt->GetRestoredSlot(methodSlot);
+    PCODE base = classMT->GetRestoredSlot(methodSlot);
+
+    if (actual == base)
+    {
+        return FALSE;
+    }
+
+    if (!classMT->IsZapped())
+    {
+        // If mscorlib is JITed, the slots can be patched and thus we need to compare the actual MethodDescs
+        // to detect match reliably
+        if (MethodTable::GetMethodDescForSlotAddress(actual) == MethodTable::GetMethodDescForSlotAddress(base))
+        {
+            return FALSE;
+        }
+    }
+
+    return TRUE;
+}
+
+static MethodTable * g_pValueTypeMT;
+static WORD g_slotEquals;
+static WORD g_slotGetHashCode;
+
+static BOOL HasOverriddenEqualsOrGetHashCodeMethodWorker(MethodTable* mt)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    } CONTRACTL_END;
+
+    _ASSERTE(mt != NULL);
+    _ASSERTE(g_pValueTypeMT != NULL);
+    _ASSERTE(g_slotEquals != 0);
+    _ASSERTE(g_slotGetHashCode != 0);
+
+    if (mt->HasCheckedOverriddenEqualsOrGetHashCode())
+    {
+        return mt->HasOverriddenEqualsOrGetHashCode();
+    }
+
+    // Check the input type.
+    if (HasOverriddenMethod(mt, g_pValueTypeMT, g_slotEquals)
+        || HasOverriddenMethod(mt, g_pValueTypeMT, g_slotGetHashCode))
+    {
+        mt->SetHasOverriddenEqualsOrGetHashCode();
+        mt->SetHasCheckedOverriddenEqualsOrGetHashCode();
+
+        // If overridden Equals or GetHashCode found, stop searching further.
+        return TRUE;
+    }
+
+    // The type itself did not override Equals or GetHashCode, go for its fields.
+    ApproxFieldDescIterator iter = ApproxFieldDescIterator(mt, ApproxFieldDescIterator::INSTANCE_FIELDS);
+    FieldDesc* pField = iter.Next();
+    while (pField != NULL
+        && pField->GetFieldType() != ELEMENT_TYPE_VALUETYPE
+        && pField->GetFieldType() != ELEMENT_TYPE_CLASS)
```

Comment:
Note that we need to check value type fields only.  The `mt->ContainsPointers()` condition in `CanUseFastGetHashCodeHelper` takes care of class reference fields.